server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -249,14 +250,23 @@ public int getNumReducePhases() {
 
         @Override
         public synchronized void close() {
-            assert hasPendingMerges() == false : "cannot close with partial reduce in-flight";
-            if (hasFailure()) {
-                assert circuitBreakerBytes == 0;
-                return;
+            try {
+                if (hasPendingMerges()) {
+                    throw new IllegalStateException("Attempted to close with partial reduce in-flight");
+                }
+                if (hasFailure()) {
+                    assert circuitBreakerBytes == 0;
+                    return;
+                }
+                assert circuitBreakerBytes >= 0;
+                circuitBreaker.addWithoutBreaking(-circuitBreakerBytes);
+                circuitBreakerBytes = 0;
+            } finally {
+                // Clean up any delayed aggregations we haven't processed.
+                Releasables.close(
+                    buffer.stream().filter(b -> b.aggregations() != null).map(b -> b.aggregations()).collect(Collectors.toList())
+                );
```

Comment:
Suggest avoiding the intermediate list altogether (unfortunately type inference isn't clever enough to do this on its own)

```suggestion
                Releasables.close(
                    (Iterable<DelayableWriteable<InternalAggregations>>) (() -> buffer.stream()
                        .map(QuerySearchResult::aggregations)
                        .iterator())
                );
```